design-decisions/2018-10-12-Condition-Split.md

```diff
@@ -0,0 +1,87 @@
+
+`rule X:Int / Y:Int => X /Int Y requires Y =/=Int 0`
+which would be written as 
+`Y =/= 0 = true /\ X:Int / Y:Int => X /Int Y`
+
+Then `3 / A:Int /\ (Y =/=Int 0 = true /\ X:Int / Y:Int)` will simplify to
+`(X:Int = 3 /\ Y:Int = A:Int) /\ (Y =/=Int 0) = true` 
+which is satisfiable, but not `T`.
+
+### Example (from unification)
+
+Consider rewriting term `if A:Int <Int 0 then -1 else 1` using
+
+`rule if true then E1:Exp else E2:Exp => E1`
+
+Then `if A:Int <Int 0 then -1 else 1 /\ if true then E1:Exp else E2:Exp` will
+simplify to `(E1:Exp = -1 /\ E2:Exp = 1) /\ A:Int <Int 0 = true`
+which is satisfiable, but not `T`.
+
+Problem
+-------
+
+Simply continuing the rewriting process when a constraint is introduced
+is not sound when the constraint does not simplify to top.  The reason is
+that, going further, the coverage is reduced for the inputs which satisfy
+the constraint.
+
+Questions
+---------
+
+* When/how we decide that a new constraint is satisfiable but not top?
+
+* How about such constraints coming from function application? 
+
+Decision: do not apply function axioms if they introduce new constraints
+------------------------------------------------------------------------
+
+We might revert this decision later, but for now this is a safe bet, as it
+prevents undesirable unrolling of functions.
+
+Decision: check whether new constraints are implied by existing ones
+--------------------------------------------------------------------
+
+To exhibit the problem described above, a new constraint must not be implied
+by the existing constraint, and their conjunction should be satisfiable but
+not top.  That is, supposing `CC` is the existing configuration constraint
+and `NC` is the new constraint, it must be that 
+* `~(CC -> NC)`, which is `CC /\ ~ NC`, is not unsatisfiable
+  (meaning `NC` is not "contained" in `CC`)
+* `CC /\ NC` is not unsatisfiable (meaning `CC /\ NC` has solutions)
+* `~(CC /\ NC)` is not unsatisfiable (meaning `CC /\ NC` is not top)
+
+Decision: split configurations whenever the above problem is exhibited
+----------------------------------------------------------------------
+
+Suppose the configuration term to be rewritten is of the form `CT /\ CC` where
+`CT` is the configuration term and `CC` is the configuration constraint.
+
+Suppose that `NC` is the new constraint generated from attempting to apply a
+rule, and that `NC` exhibits the problem above.
+
+Then, transform `CT /\ CC` succesively into the following equivalent patterns
+
+* `CT /\ CC /\ \top`
+* `CT /\ CC /\ (NC \/ \not(NC))`
+* `(CT /\ CC /\ NC) \/ (CT /\ CC /\ \not(NC))`
+
+We now have two configuration patterns. 
+
+* For the first configuration, retrying the application of the axiom should
+now succeed, as now `NC` is implied by `CC /\ NC`.
+* The second will carry the information that the inputs satisfying `NC`
+  are handled by a separate branch so they don't need to be considered
+  on this branch further.
```

Comment:
I'd add a few words about special-casing this for evaluating functions with multiple branches.

Reply:
I don't understand either. multiple overlapping step rules rules should probably lead to a case split, but I don't think we want to make a case split when a function application isn't ready to reduce.